File: src/func.py
import os
import requests
from bs4 import BeautifulSoup

def html_parser(url: str):
    """Wrapper function to minimize repeating code.

    Args:
        url (str): url

    Returns:
        BeautifulSoup: BeautifulSoup object
    """
    temp = BeautifulSoup(requests.get(url).content,"html.parser")
    return temp

def create_assets_url(url: str) -> str:
    """Creates assets url to pass into following functions.

    Args:
        url (str): url to github repo releases page.

    Returns:
        str: url of assets page for latest release tag.
    """
    #obtain latest release tag
    temp_html = html_parser(url)
    href = temp_html.find("span",class_="f1 text-bold d-inline mr-3").find("a",class_="Link--primary Link").get("href")
    tag = href.split("/")[-1]

    #create full url for expanded assets page
    expanded_assets = "https://github.com/ankidroid/Anki-Android/releases/expanded_assets/"
    full_url = f"{expanded_assets}{tag}"

    return full_url

def all_asset_urls(url: str) -> list:
    """Return list of all versions available for download.

    Args:
        url (str): url to assets page for latest release tag.

    Returns:
        list: All versions available for download as list of strings.
    """
    #parse expanded assets page
    temp_html = html_parser(url)
    apk_link = temp_html.find("div",class_="Box Box--condensed mt-3").find("ul").find_all("li")

    #store all apk urls in list
    cleanURLs = []
    for i in range(len(apk_link)):
        cleanURLs.append(apk_link[i].find("a",href=True)["href"])

    return cleanURLs

def create_metadata(url_list: list,selection: str):
    """Creates metadata to pass into download_apk function.

    Args:
        url_list (list): List of all versions available for download.
        selection (str): Simple filter to select single apk to download. Will always download 1st entry (index 0) if > 1 remains after selection.
    """
    class info:
        #filter by selection criteria (will only return single string)
        single_apk = list(filter(lambda x: selection in x,url_list))[0]

        filename = single_apk.split("/")[-1]

    return info

def download_apk(metadata):
    """Downloads apk based on passed args. Saves to Storage folder in repo.

    Args:
        metadata (class): Contains tag name and file name.
    """

    final_url = f"https://github.com{metadata.single_apk}"

    #save to Storage folder
    try:
        res = requests.get(final_url)
        with open(f"./Storage/{metadata.filename}","wb") as file:
            file.write(res.content)
        print("Success")
    except:
        print(f"Unable to save {metadata.filename} from {final_url}.")

def clear_folder(dir: str):
    """Clears all files in specified directory.

    Args:
        dir (str): Path to folder.
    """
    try:
        files = os.listdir(dir)
        for file in files:
            file_path = os.path.join(dir,file)
            if os.path.isfile(file_path):
                os.remove(file_path)
        print(f"All files in {dir} cleared.")
    except OSError:
        print(f"Error clearing all files in {dir}.")

File: src/func.rs
use reqwest::{Client,Error};
use scraper::{Html,Selector};
use std::fs::File;
use std::io;

//return html from input url
pub async fn get_html(url: &str) -> Result<Html,Error>{
    let body = Client::new()
        .get(url)
        .send()
        .await?
        .text()
        .await?;
    let raw_html = Html::parse_document(&body); //convert body to html object (parse prep)
    
    Ok(raw_html)
}

//parse latest release
pub async fn parse_tag(url: &str) -> Result<String,Box<dyn std::error::Error>>{
    //save html
    let orig_html = get_html(url).await?;

    //parse html for span tag of f1... class
    let selector = Selector::parse(
        "span.f1.text-bold.d-inline.mr-3 a.Link--primary.Link"
        )
        .unwrap();

    //append all href to vector
    let mut href_vec = Vec::new();
    for element in orig_html.select(&selector){
        if let Some(href) = element.value().attr("href"){
            href_vec.push(href.to_string());
        }
    }
    // debug_fn(&href_vec);

    //obtain latest release tag
    let latest_release = href_vec
        .first()
        .unwrap()
        .split("/")
        .last()
        .unwrap_or("")
        .to_owned();
    // debug_fn(&latest_release);

    Ok(latest_release)
}

//parse specific ver
pub async fn parse_ver(url: String,pattern: &str) -> Result<String,Box<dyn std::error::Error>>{
    //concat to get expanded_assets page
    let new_url = "https://github.com/ankidroid/Anki-Android/releases/expanded_assets/".to_owned()+url.as_str();
    // println!("{new_url}");

    //find all href
    let new_orig_html = get_html(&new_url).await?;
    let new_selector = Selector::parse(
        "div.Box.Box--condensed.mt-3 ul li a[href]"
        )
        .unwrap();

    let clean_urls: Vec<_> = new_orig_html
        .select(&new_selector)
        .filter_map(|element| element.value().attr("href"))
        .collect();
    // debug_fn(&clean_urls);

    //filter vec for parallel.A
    let href_vec_filt: Vec<_> = clean_urls
        .iter()
        .filter(|&n| n.contains(pattern))
        .cloned()
        .collect();
    // debug_fn(&href_vec_filt);

    //final url to download from
    let apk_url = "https://github.com".to_owned()+href_vec_filt[0];

    Ok(apk_url)
}

//download file
type Res<T> = std::result::Result<T,Box<dyn std::error::Error + Send + Sync>>;

pub async fn download_file(url: String,file_name: String) -> Res<()>{
    let response = reqwest::get(&url).await?;
    let mut file = File::create(file_name)?;
    
    //save file
    let mut content = io::Cursor::new(response.bytes().await?);
    io::copy(&mut content,&mut file)?;
    
    Ok(())
}

//create filename from apk_url
pub fn create_filename(arg: String) -> String{
    /*
        split by / -> index [-1] -> convert to &str -> unwrap from Option<&str>
    */
    let filename = arg.split("/").last().unwrap().to_owned();
    return filename;
}

//print resulting html output
pub fn debug_fn<T>(fn_input: &T) where T: std::fmt::Debug{
    println!("{:?}",fn_input);
}